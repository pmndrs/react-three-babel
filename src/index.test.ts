import * as babel from '@babel/core'
import { it, expect } from 'vitest'
import plugin from '../src/index'

function transform(code: string): string {
  return babel.transform(code.replace(/\s+/g, ' '), { plugins: [plugin] })!.code!
}

it('removes namespaced extend calls', () => {
  const code = transform(
    `
      import * as THREE from "three";
      import { extend } from "@react-three/fiber";

      extend(THREE);
    `,
  )
  expect(code).toMatchSnapshot()
})

it('handles JSX', () => {
  const code = transform(
    `
      function Box(props) {
        return (
          <mesh {...props}>
            <boxGeometry />
            <meshBasicMaterial />
          </mesh>
        );
      }
    `,
  )
  expect(code).toMatchSnapshot()
})

it('handles JSX member expressions', () => {
  const code = transform(
    `
      import { animated } from "@react-spring/three";

      function Box(props) {
        return (
          <animated.mesh {...props}>
            <boxGeometry />
            <meshBasicMaterial />
          </animated.mesh>
        );
      }
    `,
  )
  expect(code).toMatchSnapshot()
})

it('handles dynamic expressions', () => {
  const code = transform(
    `
      import { animated } from "@react-spring/three";

      function Material(props) {
        const Comp = props.foo ? 'meshBasicMaterial' : 'meshStandardMaterial';
        return <Comp color="royalblue" />;
      }
    `,
  )
  expect(code).toMatchSnapshot()
})

it('handles template strings', () => {
  const code = transform(
    `
      function Material(props) {
        const Comp = \`mesh\$\{props.type\}Material\`;
        return <Comp color="royalblue" />;
      }
    `,
  )
  expect(code).toMatchSnapshot()
})
