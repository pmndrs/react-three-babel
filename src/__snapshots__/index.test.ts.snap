// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handles JSX 1`] = `
"import { Mesh as _Mesh, BoxGeometry as _BoxGeometry, MeshBasicMaterial as _MeshBasicMaterial } from \\"three\\";
import { extend as _extend } from \\"@react-three/fiber\\";
_extend({
  Mesh: _Mesh,
  BoxGeometry: _BoxGeometry,
  MeshBasicMaterial: _MeshBasicMaterial
});
function Box(props) {
  return <mesh {...props}> <boxGeometry /> <meshBasicMaterial /> </mesh>;
}"
`;

exports[`handles JSX member expressions 1`] = `
"import { animated } from \\"@react-spring/three\\";
import { Mesh as _Mesh, BoxGeometry as _BoxGeometry, MeshBasicMaterial as _MeshBasicMaterial } from \\"three\\";
import { extend as _extend } from \\"@react-three/fiber\\";
_extend({
  Mesh: _Mesh,
  BoxGeometry: _BoxGeometry,
  MeshBasicMaterial: _MeshBasicMaterial
});
function Box(props) {
  return <animated.mesh {...props}> <boxGeometry /> <meshBasicMaterial /> </animated.mesh>;
}"
`;

exports[`handles dynamic expressions 1`] = `
"import { MeshBasicMaterial as _MeshBasicMaterial, MeshStandardMaterial as _MeshStandardMaterial } from \\"three\\";
import { extend as _extend } from \\"@react-three/fiber\\";
_extend({
  MeshBasicMaterial: _MeshBasicMaterial,
  MeshStandardMaterial: _MeshStandardMaterial
});
function Material(props) {
  const Comp = props.foo ? 'meshBasicMaterial' : 'meshStandardMaterial';
  return <Comp color=\\"royalblue\\" />;
}"
`;

exports[`handles template strings 1`] = `
"import { MeshBasicMaterial as _MeshBasicMaterial, MeshDepthMaterial as _MeshDepthMaterial, MeshDistanceMaterial as _MeshDistanceMaterial, MeshLambertMaterial as _MeshLambertMaterial, MeshMatcapMaterial as _MeshMatcapMaterial, MeshNormalMaterial as _MeshNormalMaterial, MeshPhongMaterial as _MeshPhongMaterial, MeshPhysicalMaterial as _MeshPhysicalMaterial, MeshStandardMaterial as _MeshStandardMaterial, MeshToonMaterial as _MeshToonMaterial } from \\"three\\";
import { extend as _extend } from \\"@react-three/fiber\\";
_extend({
  MeshBasicMaterial: _MeshBasicMaterial,
  MeshDepthMaterial: _MeshDepthMaterial,
  MeshDistanceMaterial: _MeshDistanceMaterial,
  MeshLambertMaterial: _MeshLambertMaterial,
  MeshMatcapMaterial: _MeshMatcapMaterial,
  MeshNormalMaterial: _MeshNormalMaterial,
  MeshPhongMaterial: _MeshPhongMaterial,
  MeshPhysicalMaterial: _MeshPhysicalMaterial,
  MeshStandardMaterial: _MeshStandardMaterial,
  MeshToonMaterial: _MeshToonMaterial
});
function Material(props) {
  const Comp = \`mesh\${props.type}Material\`;
  return <Comp color=\\"royalblue\\" />;
}"
`;

exports[`removes namespaced extend calls 1`] = `
"import * as THREE from \\"three\\";
import { extend } from \\"@react-three/fiber\\";"
`;
